cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/MetadataStoreDataset.java

```diff
@@ -91,11 +95,27 @@ public MetadataStoreDataset(Table table) {
   // lists all that has first id parts in range of startId and stopId
   public <T> List<T> list(Key startId, @Nullable Key stopId, Class<T> classOfT, int limit,
                           Predicate<T> filter) {
+    return Lists.newArrayList(listKV(startId, stopId, classOfT, limit, filter).values());
+  }
+
+  // returns mapping of all that has same first id parts
+  public <T> Map<byte[], T> listKV(Key id, Class<T> classOfT) {
+    return listKV(id, classOfT, Integer.MAX_VALUE);
+  }
+
+  // returns mapping of  all that has same first id parts, with a limit
+  public <T> Map<byte[], T> listKV(Key id, Class<T> classOfT, int limit) {
+    return listKV(id, null, classOfT, limit, Predicates.<T>alwaysTrue());
+  }
+
+  // returns mapping of all that has first id parts in range of startId and stopId
+  public <T> Map<byte[], T> listKV(Key startId, @Nullable Key stopId, Class<T> classOfT, int limit,
+                                   Predicate<T> filter) {
     byte[] startKey = startId.getKey();
     byte[] stopKey = stopId == null ? Bytes.stopKeyForPrefix(startKey) : stopId.getKey();
 
     try {
-      List<T> list = Lists.newArrayList();
+      Map<byte[], T> map = Maps.newLinkedHashMap();
```

Comment:
For `byte[]` as key, you cannot use a hash map. You have to use a TreeMap with a `Bytes.BYTES_COMPARATOR`